app/controllers/password_resets_controller.rb

```diff
@@ -45,11 +47,23 @@ def get_user
       @user = User.find_by(email: params[:email])
     end
     
+    def check_user
+      if !@user
+        redirect_to root_url
+      end
+    end
+
     def check_expiration
       if @user.password_reset_expired?
         flash[:danger] = "Password reset has expired."
         redirect_to new_password_reset_url
       end
     end
 
+    def check_reset_token
+      if !@user.reset_digest || BCrypt::Password.new(@user.reset_digest) != params[:id]
```

Comment:
Similarly, try unless @user.nil? && opposite of second condition.

On a second note, shouldn't it be:
`BCrypt::Password.new(params[:id])` that's being compared to `@user.reset_digest`?

Tracking the line of logic, it seems the reset_digest on the user is being set using a generated token ran through the BCrypt password creator. This token is then passed as the id to the reset password. So it would stand to reason that you use the BCrypt to resolve the id of the reset link to compare it to the stored hash in reset_digest, right?

Reply:
Counterintuitively, no. See here: https://github.com/codahale/bcrypt-ruby#how-to-use-bcrypt-ruby-in-general